### ISSUE TITLE
optimize.linprog interior-point

### ISSUE THREAD

**Issue by Eight1911** [scipy/scipy#10382]
Currently, scipy.stats.wasserstein_distance only supports distributions over the real numbers. I'm suggesting we add something like this

	wasserstein_distance(u, v, u_weights=None, v_weights=None, metric="euclidean")

where u and v can now be a matrix.

**Comment by mdhaber** [comment]
This sounds interesting. To be included in SciPy, we'd need to show that its use is not restricted to one or two specific fields, and we need an algorithm - the better known and standard the better. Can you post some more about that?

**Comment by Eight1911** [comment]
Wasserstein distance is

- the minimum distance 10 different cars need to travel to pick up 10 different passengers
- the distance between color palettes in abstract color space
- [a nice way to compute translate from word embeddings to sentence embedding](http://proceedings.mlr.press/v37/kusnerb15.pdf)

In general, it's a natural notion of distance between distributions over _any_ metric space. Wasserstein distance is also used in WGANs because it provides a nice gradient for back-propagation. (although it wasn't directly computed in that case.) 

When no weights are given and `u_values` and `v_values` have the same length, the problem becomes a minimum cost matching problem and we can use `scipy.optimize.linear_sum_assignment` to compute the distance. In general, it's a flow problem and we can use [this](https://en.wikipedia.org/wiki/Network_simplex_algorithm) to solve the flow-graph formulation.

The downside is that the distance is rather slow to compute compared to other distances.

**Comment by rgommers** [comment]
> This can be done quite easily using linear programming

That doesn't sound very healthy for `stats`. Can't this simply be done via the metrics in `scipy.spatial`?

**Comment by rlucas7** [comment]
I looked into this back in 2017 with a former intern (Lin Su), we presented findings at the optimal transport workshop at NIPS 2017 (now called NeurIPS) in Long Beach. The faster ways to do this are still linear programs with some entropy regularization which gives an interior point solution and spreads the solution around (not nice if you want a sparse representation). We looked into an L2 penalty instead which gives a sparse transport plan. Details of our study are here: https://arxiv.org/abs/1712.02512
and the full workshop proceedings materials are here: http://otml17.marcocuturi.net/schedule/
Some form Sinkhorn iteration and Greenkhorn iteration might be interesting-they are fast but approximate, see e.g. https://arxiv.org/abs/1705.09634 presented as spotlight at same conference. Although I'm not sure where would be best for something like that if your interested and if the SciPy members think that would be a good fit?

**Comment by Eight1911** [comment]
@rlucas7, thank you for your mentions.

@rgommers, Unfortunately, the metrics in `scipy.spatial.distance` do not help with computing the Wasserstein distance, which is precisely why I think this will be a valuable addition. If you'd like, we can add this to `scipy.spatial.distance` instead. Due to being more specialized, the network simplex algorithm linked should be much faster than `scipy.optimize.linprog.`

**Comment by rgommers** [comment]
Thanks @Eight1911 and @rlucas7.  Okay, so there's no shortcut:) `linear_sum_assignment` would help, or the network simplex algorithm. Is an implementation of the latter available? I'm just wondering about complexity, and if that would be generally useful and it would make sense to put that in `optimize`.

**Comment by rlucas7** [comment]
The Wasserstein distance is the solution to an optimal transport problem so somewhat of a grey area, could be a distance measure could be an optimization routine. More generally the routines would be useful for optimizations (optimal transport) maybe workhorse method in optimize and the distance function calls the workhorse?

Sincerely, 

-Lucas Roberts

**Comment by Eight1911** [comment]
@rgommers Apologies for the delay. I wrote up a bare bone implementation of the algorithm [here](https://github.com/Eight1911/network-simplex). Like the general simplex method, the algorithm is exponential time, but works very well in practice. There are more complicated polynomial time variants, but I have never read about those.

Now, that I think about it, the network simplex method can also be used to solve min-cut max-flow problems.

**Comment by rgommers** [comment]
>  I wrote up a bare bone implementation of the algorithm here. 

that link doesn't work for me. perhaps it's a private repo?

**Comment by Eight1911** [comment]
Whoops. You are right. It's now public. Needless to say, a lot of things, including more detailed comments, have to addded.

**Comment by rgommers** [comment]
Seeing it now, thanks. It's nice that you can test against `optimize.linprog`, that makes life a lot easier. 

Looking at the Cython code, the first question is if you really need all the manual memory management. Using typed memoryviews would be a lot cleaner. Usually it's as performant. 

> More generally the routines would be useful for optimizations (optimal transport) maybe workhorse method in optimize and the distance function calls the workhorse?

this approach makes sense to me

**Comment by Eight1911** [comment]
@rgommers You are right that memoryviews are cleaner. This can be changed later in the PR. Should I get started on the PR?

**Comment by rgommers** [comment]
Starting on a PR sounds good to me

**Comment by JonathanVacher** [comment]
Hi everyone,
I noticed that the current implementation of the 1d Wasserstein distance (scipy.stats.wasserstein_distance) is based on '_cdf_distance'. This is fine for now because the l_1 Wasserstein distance is equal to the Cramer distance (related to the scipy.stats.energy_distance) but for other l_p costs this is not the case.

I think it could be a good idea to separate wasserstein_distance from energy_distance and to clarify the documentation (https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html) which is restricted to l_1 Wasserstein distance and which define it as being equal to the Cramer distance. This could be misleading for a naive reader.

**Comment by morales-gregorio** [comment]
Hey all, 

I was looking for an implementation of a multi-dimensional wasserstein distance and stumbled upon this issue. Unfortunately, the link from @Eight1911 appears dead to me. 

Has there been any progress on this? Is the implementation already available in some branch? 

Thanks!

**Comment by JonathanVacher** [comment]
Hi,
You should look at the Python Optimal Transport library: https://pot.readthedocs.io/en/stable/index.html
But OT still has difficulties to handle high dimension (beyond 3/4/5 I think).

**Comment by angelotc** [comment]
Any movement on this? @Eight1911  would I be able to take a look into your code?

**Comment by mdhaber** [comment]
There is work on network simplex in gh-14540. I'll add the spatial label since it seems like that might be more appropriate. Crossref gh-7804.

**Comment by com3dian** [comment]
Hi all, 
Is there any recent update on this? I'm thinking of implementing an n-dimensional Wasserstein distance(the Earth Mover distance) in Scipy, could be a reimplementation of [R `transport`](https://www.rdocumentation.org/packages/transport/versions/0.12-2/topics/transport) or similar approach, not sure if it is still a good idea.
Thank you!

**Comment by morales-gregorio** [comment]
> Hi all, Is there any recent update on this? I'm thinking of implementing an n-dimensional Wasserstein distance(the Earth Mover distance) in Scipy, could be a reimplementation of [R `transport`](https://www.rdocumentation.org/packages/transport/versions/0.12-2/topics/transport) or similar approach, not sure if it is still a good idea. Thank you!

Hi! 

I would be interested in bringing this to scipy. 

There is also an [implementation in OpenCV called EMD](https://docs.opencv.org/3.4/d6/dc7/group__imgproc__hist.html#ga902b8e60cc7075c8947345489221e0e0) that relies on some C script. See [here](https://github.com/INM-6/NetworkUnit/blob/v0.2/networkunit/scores/wasserstein_distance.py) for how we use it in our project.

We found the openCV implementation to work fine with higher dimensions, but it gets memory issues when the sample size is very large (N > 10K observations).

**Comment by com3dian** [comment]
Hi @morales-gregorio  good to know that! Have you also found any references for the algorithm used in `cv2.EMD`, I read the doc and can't find any:)

**Comment by morales-gregorio** [comment]
Hi @com3dian,

it is a bit hidden, if you go to [the C implementation script](https://github.com/opencv/opencv/blob/8b4fa2605e1155bbef0d906bb1f272ec06d7796e/modules/imgproc/src/emd.cpp) (apparently written in 1998) it says:

>  Partially based on Yossi Rubner code:
Last update: 3/14/98
An implementation of the Earth Movers Distance. 
Based of the solution for the Transportation problem as described in "Introduction to Mathematical Programming" by F. S. Hillier and G. J. Lieberman, McGraw-Hill, 1990.

**Comment by com3dian** [comment]
@morales-gregorio thanks! will look at it soon.

**Comment by mdhaber** [comment]
Looks like this was closed by gh-17290.
